<a href="https://colab.research.google.com/github/Davis-Tony/DS-Unit-2-Applied-Modeling/blob/master/TonyDavis_LS_DS_231_assignment.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Lambda School Data Science

*Unit 2, Sprint 3, Module 1*

---


# Define ML problems

You will use your portfolio project dataset for all assignments this sprint.

## Assignment

Complete these tasks for your project, and document your decisions.

- [ ] Choose your target. Which column in your tabular dataset will you predict?
- [ ] Is your problem regression or classification?
- [ ] How is your target distributed?
    - Classification: How many classes? Are the classes imbalanced?
    - Regression: Is the target right-skewed? If so, you may want to log transform the target.
- [ ] Choose your evaluation metric(s).
    - Classification: Is your majority class frequency >= 50% and < 70% ? If so, you can just use accuracy if you want. Outside that range, accuracy could be misleading. What evaluation metric will you choose, in addition to or instead of accuracy?
    - Regression: Will you use mean absolute error, root mean squared error, R^2, or other regression metrics?
- [ ] Choose which observations you will use to train, validate, and test your model.
    - Are some observations outliers? Will you exclude them?
    - Will you do a random split or a time-based split?
- [ ] Begin to clean and explore your data.
- [ ] Begin to choose which features, if any, to exclude. Would some features "leak" future information?

If you haven't found a dataset yet, do that today. [Review requirements for your portfolio project](https://lambdaschool.github.io/ds/unit2) and choose your dataset.

Some students worry, ***what if my model isn't “good”?*** Then, [produce a detailed tribute to your wrongness. That is science!](https://twitter.com/nathanwpyle/status/1176860147223867393)

In [1]:
import sys
!pip install category_encoders==2.*

In [2]:
#Reading in data
import pandas as pd

url='https://raw.githubusercontent.com/Davis-Tony/DS-Unit-2-Applied-Modeling/master/WhoopData.csv'

df = pd.read_csv(url)

In [3]:
df.head()

date  ...  whoop_nap_duration_total_asleep
0  2020-02-21  ...                              NaN
1  2020-02-22  ...                              NaN
2  2020-02-23  ...                              NaN
3  2020-02-24  ...                              NaN
4  2020-02-25  ...                              NaN

[5 rows x 32 columns]

In [5]:
#Beginning data explortaion
#Checking for missing data
df.isnull().sum()

date                                    0
whoop_strain_maxhr                      0
whoop_strain_score                      0
whoop_workout_count                   121
whoop_strain_avghr                      0
whoop_strain_calories                   0
whoop_workout_duration                121
whoop_sleep_awake_percent               1
whoop_sleep_disturbances_per_hour       1
whoop_sleep_duration_deep_percent       1
whoop_sleep_duration_rem_percent        1
whoop_recovery_rhr                      1
whoop_recovery_hrv                      1
whoop_sleep_duration_light              1
whoop_sleep_start_standardized          1
whoop_sleep_duration_awake              1
whoop_sleep_duration_total_inbed        1
whoop_sleep_duration_deep               1
whoop_sleep_disturbances                1
whoop_sleep_duration_light_percent      1
whoop_sleep_cycles                      1
whoop_sleep_duration_rem                1
whoop_sleep_asleep_percent              1
whoop_sleep_duration_total_asleep 

In [6]:
#dropping nap columns as they're nearly totally empty
df = df.drop(['whoop_nap_count', 'whoop_nap_duration_total_asleep'], axis=1)

In [7]:
df.head()

date  ...  whoop_sleep_score_consistency
0  2020-02-21  ...                            NaN
1  2020-02-22  ...                            NaN
2  2020-02-23  ...                            NaN
3  2020-02-24  ...                            NaN
4  2020-02-25  ...                           59.0

[5 rows x 30 columns]

In [8]:
#Dropping first row as most of the info included is NaN
df = df.drop([0])

In [9]:
df.head()

date  ...  whoop_sleep_score_consistency
1  2020-02-22  ...                            NaN
2  2020-02-23  ...                            NaN
3  2020-02-24  ...                            NaN
4  2020-02-25  ...                           59.0
5  2020-02-26  ...                           69.0

[5 rows x 30 columns]

In [10]:
df.dtypes

date                                   object
whoop_strain_maxhr                      int64
whoop_strain_score                    float64
whoop_workout_count                   float64
whoop_strain_avghr                      int64
whoop_strain_calories                 float64
whoop_workout_duration                float64
whoop_sleep_awake_percent             float64
whoop_sleep_disturbances_per_hour     float64
whoop_sleep_duration_deep_percent     float64
whoop_sleep_duration_rem_percent      float64
whoop_recovery_rhr                    float64
whoop_recovery_hrv                    float64
whoop_sleep_duration_light            float64
whoop_sleep_start_standardized        float64
whoop_sleep_duration_awake            float64
whoop_sleep_duration_total_inbed      float64
whoop_sleep_duration_deep             float64
whoop_sleep_disturbances              float64
whoop_sleep_duration_light_percent    float64
whoop_sleep_cycles                    float64
whoop_sleep_duration_rem          

In [11]:
#Converting date to datetime format
df['date'] = pd.to_datetime(df['date'])

In [12]:
df.dtypes

date                                  datetime64[ns]
whoop_strain_maxhr                             int64
whoop_strain_score                           float64
whoop_workout_count                          float64
whoop_strain_avghr                             int64
whoop_strain_calories                        float64
whoop_workout_duration                       float64
whoop_sleep_awake_percent                    float64
whoop_sleep_disturbances_per_hour            float64
whoop_sleep_duration_deep_percent            float64
whoop_sleep_duration_rem_percent             float64
whoop_recovery_rhr                           float64
whoop_recovery_hrv                           float64
whoop_sleep_duration_light                   float64
whoop_sleep_start_standardized               float64
whoop_sleep_duration_awake                   float64
whoop_sleep_duration_total_inbed             float64
whoop_sleep_duration_deep                    float64
whoop_sleep_disturbances                     f

In [13]:
df.head()

date  ...  whoop_sleep_score_consistency
1 2020-02-22  ...                            NaN
2 2020-02-23  ...                            NaN
3 2020-02-24  ...                            NaN
4 2020-02-25  ...                           59.0
5 2020-02-26  ...                           69.0

[5 rows x 30 columns]

In [14]:
#Creating new feature to qualify great sleep
df['great_recovery'] = df['whoop_recovery_score'] > 66

In [15]:
#Verifying
df.head()

date  whoop_strain_maxhr  ...  whoop_sleep_score_consistency  great_recovery
1 2020-02-22                 148  ...                            NaN           False
2 2020-02-23                 126  ...                            NaN           False
3 2020-02-24                 180  ...                            NaN           False
4 2020-02-25                 178  ...                           59.0           False
5 2020-02-26                 165  ...                           69.0            True

[5 rows x 31 columns]

In [16]:
df['great_recovery'].value_counts(normalize=True)

False    0.621622
True     0.378378
Name: great_recovery, dtype: float64

This will be a binary classification problem as the target will be the 'great_recovery' column split in two classes. With the majority class at 62%, this is a pretty balanced split.

Since it's balanced, we'll use accuracy as the evaluation metric.

I'll be using a time-based split using a specific date as the split between test and train datasets.

The one feature that would result in leakage is the 'whoop_recovery_score' column beacuse that directly feeds the targeted 'great_recovery' column.